### PR TITLE
fix: consistency in filepaths

### PR DESCRIPTION
**Pull request by Fersoil**
Unified the approach for saving files in `process_plate` and `generate_plate_report`. Now the file saving works following the rmarkdown::render approach and the specified guidelines. We now accept the output directory where the file should be saved and the filename.

Since the new arguments break the backwards compatibility, I raised the package version to 0.1.0

outputs of the normalisation are saved by default in `normalised_data` directory

**During review, please focus mainly on comprehensiveness of the docs regarding filepaths**

### CODE REVIEW

File: R/process-plate.R [hunk 1]
```diff
@@ -60,21 +71,26 @@ is_valid_normalisation_type <- function(normalisation_type) {
...
+#'
+#' # process plate without adjusting blanks and save the output to a file with a custom name
+#' process_plate(plate, filename = "plate_without_blanks_adjusted.csv",
+#'    output_dir = tmp_dir, adjust_blanks = FALSE)
+#'
 #'
 #' # nMFI normalisation
 #' process_plate(plate,
-#'   output_path = temporary_filepath,
+#'   output_dir = tmp_dir,
```
Comment (nizwant) [inline R/process-plate.R:84]: while I personaly do not mind `tmp_dir` I would call it in the code that is exposed to users i.e. docs to `example_directory` or something along those lines

File: R/process-plate.R [hunk 2]
```diff
@@ -34,9 +33,21 @@ is_valid_normalisation_type <- function(normalisation_type) {
...
-#' Where the `{plate_name}` is the name of the plate.
+#' @param filename (`character(1)`) The name of the output CSV file with normalised MFI values.
+#' If not provided or equals to `NULL`, the output filename will be based on the normalisation type
+#' and the plate name, precisely: `{plate_name}_{normalisation_type}.csv`.
+#' By default the `plate_name` is the filename of the input file that contains the plate data.
+#' For more details please refer to \link[PvSTATEM]{Plate}.
+#'
+#' If the passed filename does not contain `.csv` extension, the default extension `.csv` will be added.
+#' Filename can also be a path to a file, e.g. `path/to/file.csv`. In this case, the `output_dir` and `filename` will be joined together.
+#' However, if the passed filepath is a an absolute path and the `output_dir` parameter is also provided, the `output_dir` parameter will be ignored.
```
Comment (nizwant) [inline R/process-plate.R:44]: is a an absolute -> is an absolute

File: R/process-plate.R [hunk 3]
```diff
@@ -34,9 +33,21 @@ is_valid_normalisation_type <- function(normalisation_type) {
...
+#' If there already exists a file under a specified filepath, the function will overwrite it.
```
Comment (nizwant) [inline R/process-plate.R:45]: If there already exists a file under -> If a file already exists under

File: R/generate_report.R [hunk 4]
```diff
@@ -6,8 +6,20 @@
...
-#' @param output_dir (`character(1)`) The directory where the report should be saved. Default is 'reports'.
+#'
+#' @param filename (`character(1)`) The name of the output HTML report file.
+#' If not provided or equals to `NULL`, the output filename will be based on the plate name, precisely: `{plate_name}_report.html`.
+#' By default the `plate_name` is the filename of the input file that contains the plate data.
+#' For more details please refer to \link[PvSTATEM]{Plate}.
+#'
+#' If the passed filename does not contain `.html` extension, the default extension `.html` will be added.
+#' Filename can also be a path to a file, e.g. `path/to/file.html`. In this case, the `output_dir` and `filename` will be joined together.
+#' However, if the passed filepath is a an absolute path and the `output_dir` parameter is also provided, the `output_dir` parameter will be ignored.
```
Comment (nizwant) [inline R/generate_report.R:17]: same as bellow

File: R/generate_report.R [hunk 5]
```diff
@@ -6,8 +6,20 @@
...
+#' If there already exists a file under a specified filepath, the function will overwrite it.
```
Comment (nizwant) [inline R/generate_report.R:18]: same as bellow

File: R/generate_report.R [hunk 6]
```diff
@@ -6,8 +6,20 @@
...
+#'
+#' @param output_dir (`character(1)`) The directory where the output CSV file should be saved.
+#' Please note that any directory path provided will create any necessary directories if they do not exist.
+#' If equals to `NULL` the current working directory will be used. Default is 'reports'.
```
Comment (nizwant) [inline R/generate_report.R:22]: if equals -> if it equals

File: R/helpers.R [hunk 7]
```diff
@@ -225,3 +225,77 @@ is.decreasing <- function(x) {
   }
   all(diff(x) < 0)
 }
+
+
+
+#' @title Validate filepath and output_dir
+#' @description This function handles the validation of the filepath and output_dir arguments.
```
Comment (nizwant) [inline R/helpers.R:232]: handles the validation of  -> validates

**Review by nizwant — CHANGES_REQUESTED** [review]

File: R/helpers.R [hunk 8]
```diff
@@ -225,3 +225,77 @@ is.decreasing <- function(x) {
...
+  output_path <- file.path(output_dir, filename)
+
+
+  # create the directories and check if the file exists
+  output_dir <- dirname(output_path)
+  filename <- basename(output_path)
+
+  if (!dir.exists(output_dir)) {
+    verbose_cat("Creating the output directory: '", output_dir, "'\n", verbose = verbose)
+    dir.create(output_dir)
```
Comment (ZetrextJG) [inline R/helpers.R:293]: Quoting the R base docs:
```
recursive
logical. Should elements of the path other than the last be created? If true, like the Unix command mkdir -p.
Default FALSE
```
Which means that the docs are not consistent with what the program does

File: R/process-plate.R [hunk 9]
```diff
@@ -34,9 +33,21 @@ is_valid_normalisation_type <- function(normalisation_type) {
...
+#'
+#' @param output_dir (`character(1)`) The directory where the output CSV file should be saved.
+#' Please note that any directory path provided will create any necessary directories if they do not exist.
```
Comment (ZetrextJG) [inline R/process-plate.R:48]: Same as above

**Review by ZetrextJG — CHANGES_REQUESTED** [review]

File: R/process-plate.R (hunk 1 above)
Comment (Fersoil) [reply R/process-plate.R:84]: what do you mean exactly? Do you want to print the `tmp_dir` path to the user? 
Or maybe you suggest renaming this variable to `example_directory`?

File: R/helpers.R (hunk 8 above)
Comment (Fersoil) [reply R/helpers.R:293]: adjusted the `create.dir` parameters to match the described behaviour

File: R/process-plate.R (hunk 1 above)
Comment (nizwant) [reply R/process-plate.R:84]: I suggest renaming

**Review by nizwant — APPROVED** [review]

File: R/process-plate.R (hunk 1 above)
Comment (Fersoil) [reply R/process-plate.R]: done

**Pull request merged**